src/network/network_server.cpp

```diff
@@ -953,24 +965,38 @@ NetworkRecvStatus ServerNetworkGameSocketHandler::Receive_CLIENT_IDENTIFY(Packet
 	return this->SendNewGRFCheck();
 }
 
-NetworkRecvStatus ServerNetworkGameSocketHandler::Receive_CLIENT_GAME_PASSWORD(Packet &p)
+NetworkRecvStatus ServerNetworkGameSocketHandler::Receive_CLIENT_AUTH_RESPONSE(Packet &p)
 {
 	if (this->status != STATUS_AUTH_GAME) {
 		return this->SendError(NETWORK_ERROR_NOT_EXPECTED);
 	}
 
-	Debug(net, 9, "client[{}] Receive_CLIENT_GAME_PASSWORD()", this->client_id);
+	Debug(net, 9, "client[{}] Receive_CLIENT_AUTH_RESPONSE()", this->client_id);
 
-	std::string password = p.Recv_string(NETWORK_PASSWORD_LENGTH);
+	switch (this->authentication_handler->ReceiveResponse(p)) {
+		case NetworkAuthenticationServerHandler::AUTHENTICATED:
+			break;
 
-	/* Check game password. Allow joining if we cleared the password meanwhile */
-	if (!_settings_client.network.server_password.empty() &&
-			_settings_client.network.server_password.compare(password) != 0) {
-		/* Password is invalid */
-		return this->SendError(NETWORK_ERROR_WRONG_PASSWORD);
+		case NetworkAuthenticationServerHandler::RETRY_NEXT_METHOD:
+			return this->SendAuthRequest();
+
+		case NetworkAuthenticationServerHandler::NOT_AUTHENTICATED:
+		default:
+			return this->SendError(NETWORK_ERROR_WRONG_PASSWORD);
```

Comment:
Did some testing, and the one thing that stood out, was this error.

When I set a password on a server, and I enter the wrong one, this is as expected.
When I set a password on a server and add a authorized-keys, and I have the wrong client-public-key and I enter the wrong password, this is as expected.

But, when I only have authorized-keys and set no password, and I have the wrong client-public-key, I also get "wrong password". But I never got the chance to enter a password. So that feels confusing.

Could we add a new error here, something like `NETWORK_ERROR_NO_AUTHORIZED_KEYS` or `NETWORK_ERROR_NOT_ON_ALLOWLIST`, with a string representation on the client for that? So it is a bit more clear that there never was an opportunity to enter a password.